File: src/E042.py
from math import sqrt

def load():
    with open('E042.txt', 'r') as f: return eval(f.read())

def is_t(n):
    x = (sqrt(8*n+1)-1)/2
    return x-int(x) <= 0

def code(w):
    return sum([ord(c)-ord('A')+1 for c in w])

def f(data):
    n = 0
    for w in data:
        if is_t(code(w)):
            n += 1
    return n

assert(is_t(code('SKY')))
print f(load())

File: src/E042.hs
module E042 (e42) where

import Data.Char (ord)
import Data.List.Split (splitOn)
import Util (isInt)

isTriangular :: Integral a => a -> Bool
isTriangular n = isInt x
    where x = (sqrt $ fromIntegral (8*n+1)-1) / 2

code :: [Char] -> Int
code = sum . map (subtract 64 . ord)

parse :: [Char] -> [[Char]]
parse = splitOn "," . filter (/= '"')

e42 :: String -> Integer
e42 = fromIntegral . length . filter (isTriangular . code) . parse
